packages/aws-cdk/lib/api/hotswap/ecs-services.ts

```diff
@@ -167,13 +189,49 @@ export async function isHotswappableEcsServiceChange(
   return ret;
 }
 
+function mergeTaskDefinitions(patch: {containerDefinitions: {[key:string]: any}[]}, target: AWS.ECS.DescribeTaskDefinitionResponse) {
+  const src = patch.containerDefinitions;
+  const dst = target.taskDefinition?.containerDefinitions;
+  if (dst === undefined) {
+    return;
+  }
+  // schema: https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html
+  for (const i in src) {
+    for (const key of Object.keys(src[i])) {
+      (dst[i] as any)[key] = src[i][key];
+    }
+  }
+
+  // The describeTaskDefinition response contains several keys that must not exist in a registerTaskDefinition request.
+  // We remove these keys here, comparing these two structs:
+  // https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RegisterTaskDefinition.html#API_RegisterTaskDefinition_RequestSyntax
+  // https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_DescribeTaskDefinition.html#API_DescribeTaskDefinition_ResponseSyntax
+  [
+    'compatibilities',
+    'taskDefinitionArn',
+    'revision',
+    'status',
+    'requiresAttributes',
+    'compatibilities',
+    'registeredAt',
+    'registeredBy',
+  ].forEach(key=> delete (target.taskDefinition as any)[key]);
+
+  if (target.tags !== undefined && target.tags.length > 0) {
+    // the tags field is in a different location in describeTaskDefinition response, moving it as intended for registerTaskDefinition request.
+    (target.taskDefinition as any).tags = target.tags;
+  }
+}
+
 interface EcsService {
   readonly serviceArn: string;
 }
 
 async function prepareTaskDefinitionChange(
-  evaluateCfnTemplate: EvaluateCloudFormationTemplate, logicalId: string, change: HotswappableChangeCandidate,
-) {
+  evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+  logicalId: string,
+  change: HotswappableChangeCandidate,
+): Promise<undefined | { taskDefinition: any; family: string; hotswappableWithMerge: boolean }> {
```

Comment:
Can you define a named type for this, and add comments to the fields?

Also, `any`s scare me. Can we type the taskDefinition a bit more thoroughly?

Reply:
I added a named type.

I agree that `any` should be avoided, but actually it seems difficult in this case, because we are using a returned value from `evaluateCfnExpression` and its signature forces `any` as the returning type.

https://github.com/aws/aws-cdk/blob/64f2c20310ba09673c37d3d6d8700f730b957bf0/packages/aws-cdk/lib/api/hotswap/ecs-services.ts#L260-L263